src/core/SampleLoader.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * SampleLoader.cpp - Sample loader with an optional cache
+ *
+ * Copyright (c) 2023 saker <sakertooth@gmail.com>
+ *               2024 Dalton Messmer <messmer.dalton/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleLoader.h"
+
+#include <QMessageBox>
+#include <cassert>
+
+#include "Engine.h"
+#include "GuiApplication.h"
+#include "PathUtil.h"
+#include "Song.h"
+#include "lmms_basics.h"
+
+namespace lmms {
+
+namespace {
+
+void displayError(const QString& message)
+{
+	Engine::getSong()->collectError(QString("%1: %2").arg(QObject::tr("Error loading sample"), message));
+}
+
+} // namespace
+
+/**
+ * An evictor which automatically evicts sample buffers from the cache
+ * immediately after the last reference to the buffer is dropped
+*/
+class SampleLoader::AutoEvictor
+{
+public:
+	void operator()(SampleBuffer* p) const noexcept
+	{
+		if (!p) { return; }
+
+		SampleLoader::inst().remove(*p);
+		delete p;
+	}
+};
+
+SampleLoader::SampleLoader()
+{
+	connect(&m_watcher, &QFileSystemWatcher::fileChanged, this, &SampleLoader::removeFile);
+}
+
+auto SampleLoader::inst() -> SampleLoader&
+{
+	static SampleLoader singleton;
+	return singleton;
+}
+
+auto SampleLoader::fromFile(const QString& filePath, Cache cache)
+	-> std::shared_ptr<const SampleBuffer>
+{
+	if (filePath.isEmpty()) { return nullptr; }
+
+	switch (cache)
+	{
+		case Cache::Read: [[fallthrough]];
+		case Cache::ReadWrite:
+		{
+			if (auto buffer = SampleLoader::inst().getFile(filePath))
```

Comment:
Since this function is in the same scope as `SampleLoader`, I believe we can simplify the function call to just use `inst()` (the same  suggestion applies elsewhere).